### PR TITLE
React router v6

### PR DESCRIPTION
**Pull request by KutnerUri**
## Proposed Changes

- upgrade react router to v6
- upgrade all routes and links to new syntax
- [x] fix parent sync
- [x] update symphony links to v6 (see - https://github.com/teambit/symphony/pull/744)
- [x] symphony base-url solution
- [x] verify compatibility with current symphony!

## new syntax!
- RR6 Routes now use `<Route element={<div></div>}/>` instead of `<Route><div></div></Route>`
- children can only be other routes
- new feature: all routes and links are _**RELATIVE**_.

for example:
```tsx
<Routes>
  <Route path="hello" element={<div>I will render at "/hello"</div>} >
      <Route path="world" element={<div>I will render at "/hello/world"</div>} />
      <Route path="doc" element={<HelloDocPage/>} />
  </Route>
</Routes>

function HelloDocPage(){
  return <div>
    <Link to="./thanks">I will point to "/hello/doc/thanks"</Link>
    <Link to="../world">I will point to "/hello/world"</Link>
    <Link to="/hello">I am not relative and will point to "/hello"</Link>
  </div>
}
```

in addition, React Router removed the "path-to-regex" module, and now forces us to implement everything.
> React Router v6 uses a simplified path format. <Route path> in v6 supports only 2 kinds of placeholders: dynamic :id-style > params and * wildcards. A * wildcard may be used only at the end of a path, not in the middle.
> 
> All of the following are valid route paths in v6:
>
> ``` 
> /groups
> /groups/admin
> /users/:id
>/users/:id/messages
>/files/*
> /files/:id/*
> ```
> The following RegExp-style route paths are not valid in v6:
> ```
> /users/:id?
> /tweets/:id(\d+)
> /files/*/cat.jpg
> /files-*
> ```

so this type of routes will no longer work:
```
<Route path="/:compoentPage(.*)/~code/:file(.*)?" element={<CodePage/>} >
```

instead we'll have to do a complicated check, like this:
```tsx

<Route path="*" element={<ComponentPage/>} />

const parentRegex = /([^~]*[^~/])(\/~)?/;
function ComponentPage() {
    const params = useParams();
    const splat = params["*"];

    const match = splat ? parentRegex.exec(splat) : null;
    const parentPath = match ? match[1] : "/";

    return (
        <Routes>
            <Route path={`${parentPath}/*`}>
                <Route index element={<ProjectPage id={parentPath} />} />
                <Route
                    path="~code/*"
                    element={<CodePage projectId={parentPath} />}
                />
            </Route>
        </Routes>
    );
}
```

### CODE REVIEW

File: .bitmap [hunk 1]
```diff
@@ -1500,30 +1500,12 @@
         "mainFile": "index.ts",
         "rootDir": "scopes/code/queries/get-file-content"
     },
-    "ui/react-router/extend-path": {
```
Comment (KutnerUri) [inline .bitmap:1491]: ejected no longer needed comps

File: scopes/code/hooks/use-code-params/use-code-params.ts [hunk 2]
```diff
@@ -2,8 +2,8 @@ import { useParams } from 'react-router-dom';
 import { useQuery } from '@teambit/ui-foundation.ui.react-router.use-query';
 
 export type CodeRouteParams = {
-  componentId: string;
-  file: string | undefined;
+  componentId?: string;
+  file?: string | undefined;
```
Comment (KutnerUri) [inline scopes/code/hooks/use-code-params/use-code-params.ts:6]: fix types

File: scopes/code/hooks/use-code-params/use-code-params.ts [hunk 3]
```diff
@@ -12,6 +12,8 @@ export function useCodeParams(): CodeRouteParams {
   // I think we should have a hook that keeps all qury parms and just overrides a specific one
   const query = useQuery();
   const version = query.get('version') || undefined;
-  const codeRouteParams: CodeRouteParams = useParams();
-  return { ...codeRouteParams, version };
+  const codeRouteParams = useParams<CodeRouteParams>();
+  const file = codeRouteParams['*'];
```
Comment (KutnerUri) [inline scopes/code/hooks/use-code-params/use-code-params.ts:16]: new syntax - single splat

File: scopes/code/ui/code-tab-page/code-tab-page.tsx [hunk 4]
```diff
@@ -19,15 +19,11 @@ type CodePageProps = {
...
+  const urlParams = useCodeParams();
   const component = useContext(ComponentContext);
   const { mainFile, fileTree = [], dependencies, devFiles } = useCode(component.id);
 
-  const fileFromUrl = useMemo(() => {
-    return fileName.file;
-  }, [fileName]);
-
-  const currentFile = fileFromUrl || mainFile;
+  const currentFile = urlParams.file || mainFile;
```
Comment (KutnerUri) [inline scopes/code/ui/code-tab-page/code-tab-page.tsx:26]: no need for memo - just string

File: scopes/code/ui/code-tab-tree/code-tab-tree.tsx [hunk 5]
```diff
@@ -48,18 +48,16 @@ export function CodeTabTree({
...
+      const fileUrl = `${props.node.id}${affix('?version=', urlParams.version)}`;
+      // (for reference - absolute url)
+      // const href = `${currentLaneUrl}${toPathname(component.id)}/~code/${fileUrl}`
+
       const widgets = getWidgets(props.node.id, mainFile, devFiles);
       if (!children) {
         return (
           <Node
-            href={href}
+            href={`./${fileUrl}`}
```
Comment (KutnerUri) [inline scopes/code/ui/code-tab-tree/code-tab-tree.tsx:60]: now using relative link no need to be aware of parent routes

File: scopes/component/component/component.ui.runtime.tsx [hunk 6]
```diff
@@ -271,15 +269,16 @@ export class ComponentUI {
       commandBarUI,
       reactRouterUI
     );
-    const section = new AspectSection();
+    const aspectSection = new AspectSection();
```
Comment (KutnerUri) [inline scopes/component/component/component.ui.runtime.tsx:272]: just rename

File: scopes/component/component/ui/menu/menu.tsx [hunk 7]
```diff
@@ -39,24 +41,39 @@ export type MenuProps = {
 /**
  * top bar menu.
  */
-export function Menu({ navigationSlot, widgetSlot, className, host, menuItemSlot, consumeMethodSlot }: MenuProps) {
-  const { component } = useComponent(host);
+export function ComponentMenu({
```
Comment (KutnerUri) [inline scopes/component/component/ui/menu/menu.tsx:44]: rename to make the menu clearer

File: scopes/component/component/ui/menu/mobile-menu-nav.tsx [hunk 8]
```diff
@@ -27,7 +25,7 @@ export function MobileMenuNav({
   return (
     <Dropdown
       // @ts-ignore - mismatch between @types/react
-      placeholder={<Placeholder slots={totalSlots} baseUrl={url} />}
+      placeholder={<Placeholder slots={totalSlots} />}
```
Comment (KutnerUri) [inline scopes/component/component/ui/menu/mobile-menu-nav.tsx:28]: relative routes - no need for parent path

File: scopes/component/ui/version-dropdown/version-dropdown.composition.tsx [hunk 9]
```diff
@@ -15,13 +14,12 @@ export const VersionDropdownWithOneVersion = () => {
...
 export const VersionDropdownWithMultipleVersions = () => {
-  const history = createBrowserHistory();
   const versions = ['0.3', '0.2', '0.1'].map((version) => ({ version }));
   return (
     <ThemeCompositions style={style}>
-      <Router history={history}>
+      <MemoryRouter>
         <VersionDropdown tags={versions} currentVersion={versions[0].version} />
-      </Router>
+      </MemoryRouter>
```
Comment (KutnerUri) [inline scopes/component/ui/version-dropdown/version-dropdown.composition.tsx:22]: composition file - simple memory router will do

File: scopes/lanes/ui/hooks/lanes-context/lanes-model.ts [hunk 10]
```diff
@@ -67,39 +67,32 @@ export type LanesModelProps = {
...
-    // attach the base lane route
-    path = `${this.baseLaneRoute}${path}`;
-
-    const matches = LanesModel.regexp.exec(path);
+  static getLaneIdFromPathname = (pathname: string): string | undefined => {
+    const matches = LanesModel.laneFromPathRegex.exec(pathname);
     if (!matches) return undefined;
-    const [, orgId, scopeId, laneId] = matches;
-    return `${orgId ? orgId.concat('.').concat(scopeId) : scopeId}/${laneId}`;
+    const [, /* orgId, */ scopeId, laneId] = matches;
```
Comment (KutnerUri) [inline scopes/lanes/ui/hooks/lanes-context/lanes-model.ts:89]: can't have orgs in open source bit

File: scopes/lanes/ui/readme/lane-readme-overview.tsx [hunk 11]
```diff
@@ -18,10 +18,5 @@ export function LaneReadmeOverview({ host, overviewSlot, routeSlot }: LaneReadme
     return <LaneReadme host={host} viewedLane={viewedLane} overviewSlot={overviewSlot} routeSlot={routeSlot} />;
   }
 
-  if (viewedLane) {
-    return <ReactRouter.Redirect to={`${LanesModel.getLaneUrl(viewedLane.id)}/~gallery`} />;
-  }
-
-  // throw 404
-  return null;
+  return <Navigate to={`./~gallery`} />;
```
Comment (KutnerUri) [inline scopes/lanes/ui/readme/lane-readme-overview.tsx:21]: always redirect to `~gallery`, worse case the gallery will show 404

File: scopes/ui-foundation/react-router/react-router/LocationHooks.tsx [hunk 12]
```diff
@@ -0,0 +1,24 @@
+import { useEffect } from 'react';
+import { useLocation, useNavigate, NavigateFunction, useNavigationType } from 'react-router-dom';
+import { LocationListener } from './react-router.ui.runtime';
+
+export function LocationHooks({
```
Comment (KutnerUri) [inline scopes/ui-foundation/react-router/react-router/LocationHooks.tsx:5]: v6 location listener

File: scopes/ui-foundation/react-router/react-router/legacy-routing.tsx [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import React, { PropsWithChildren, forwardRef } from 'react';
+import { RoutingProvider, Routing } from '@teambit/base-ui.routing.routing-provider';
+import type { NavLinkProps } from '@teambit/base-ui.routing.native-nav-link';
+import { Link, useLocation } from '@teambit/base-react.navigation.link';
+
```
Comment (KutnerUri) [inline scopes/ui-foundation/react-router/react-router/legacy-routing.tsx:5]: adapter for old links (e.g. in menu item)

File: scopes/ui-foundation/react-router/react-router/route-context.tsx [hunk 14]
```diff
@@ -23,17 +23,15 @@ type RootRouteProps = {
 /**
  * Setup context needed for routing.
  */
-export function RouteContext({ reactRouterUi, routing = Routing.url, children, location }: RouterContextProps) {
+export function RouteContext({ routing = Routing.url, children, location }: RouterContextProps) {
   return (
-    // {/* set up the virtual router (browser, inMemory, etc) */}
     <Router type={routing} location={location}>
-      {/* injects History object back to reactRouterUi */}
-      <HistoryGetter onRouterChange={reactRouterUi.setRouter} />
```
Comment (KutnerUri) [inline scopes/ui-foundation/react-router/react-router/route-context.tsx:31]: moved history getter logic to `react-router.ui.runtime.tsx`

File: scopes/lanes/lanes/lanes.ui.runtime.tsx [hunk 15]
```diff
@@ -97,7 +110,7 @@ export class LanesUI {
     this.registerNavigation([
       {
         props: {
-          href: '',
+          href: '.',
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/lanes/lanes/lanes.ui.runtime.tsx:112]: done

**Review by luvkapur — APPROVED** [review]

File: scopes/component/component/ui/use-component.tsx [hunk 16]
```diff
@@ -13,18 +12,13 @@ export type Component = {
...
 
-type ComponentRoute = {
-  componentId?: string;
-};
-
-export function useComponent(host: string, id?: ComponentID): Component {
-  const {
-    params: { componentId },
-  } = useRouteMatch<ComponentRoute>();
+export function useComponent(host: string, id?: string | ComponentID): Component {
```
Comment (KutnerUri) [inline scopes/component/component/ui/use-component.tsx:15]: to fix
- [ ]should only get a componentID, and use its version if available, else use from params

**Pull request merged**